File: src/743_network_delay_time/network_delay_time.cpp
#include <climits>
#include <iostream>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
   public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<pair<int, int>>> adj(n + 1);
        for (auto& t : times) {
            int u = t[0], v = t[1], w = t[2];
            adj[u].push_back({v, w});
        }
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({0, k});
        unordered_set<int> visited;
        int maxTime = 0;
        while (!pq.empty()) {
            auto [dist, u] = pq.top();
            pq.pop();
            if (visited.count(u)) continue;
            visited.insert(u);
            maxTime = dist;
            for (auto& p : adj[u]) {
                int v = p.first, w = p.second;
                if (visited.count(v) == 0) {
                    pq.push({dist + w, v});
                }
            }
        }
        return visited.size() == n ? maxTime : -1;
    }
};

// Test helper
void testNetworkDelayTime() {
    Solution solution;

    // Test case 1: Example 1
    vector<vector<int>> times1 = {{2, 1, 1}, {2, 3, 1}, {3, 4, 1}};
    int n1 = 4;
    int k1 = 2;
    int expected1 = 2;
    int actual1 = solution.networkDelayTime(times1, n1, k1);
    cout << actual1 << " | " << expected1 << endl;

    // Test case 2: Example 2
    vector<vector<int>> times2 = {{1, 2, 1}};
    int n2 = 2;
    int k2 = 1;
    int expected2 = 1;
    int actual2 = solution.networkDelayTime(times2, n2, k2);
    cout << actual2 << " | " << expected2 << endl;

    // Test case 3: Example 3
    vector<vector<int>> times3 = {{1, 2, 1}};
    int n3 = 2;
    int k3 = 2;
    int expected3 = -1;
    int actual3 = solution.networkDelayTime(times3, n3, k3);
    cout << actual3 << " | " << expected3 << endl;

    // Test case 4: Single node
    vector<vector<int>> times4 = {};
    int n4 = 1;
    int k4 = 1;
    int expected4 = 0;
    int actual4 = solution.networkDelayTime(times4, n4, k4);
    cout << actual4 << " | " << expected4 << endl;

    // Test case 5: Disconnected nodes
    vector<vector<int>> times5 = {{1, 2, 1}, {3, 4, 1}};
    int n5 = 4;
    int k5 = 1;
    int expected5 = -1;
    int actual5 = solution.networkDelayTime(times5, n5, k5);
    cout << actual5 << " | " << expected5 << endl;

    // Test case 6: All connected
    vector<vector<int>> times6 = {{1, 2, 1}, {1, 3, 2}, {2, 3, 1}, {3, 4, 1}};
    int n6 = 4;
    int k6 = 1;
    int expected6 = 3;
    int actual6 = solution.networkDelayTime(times6, n6, k6);
    cout << actual6 << " | " << expected6 << endl;
}

int main() {
    testNetworkDelayTime();
    return 0;
}

File: src/743_network_delay_time/network_delay_time.py
from typing import List
import heapq


class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adjList = [[] for _ in range(n)]
        for u, v, w in times:
            adjList[u - 1].append([v, w])

        time = 0
        h = []
        heapq.heappush(h, [0, k])
        visit = set()

        while h:
            w, u = heapq.heappop(h)

            if u in visit:
                continue
            time = w
            visit.add(u)

            for v, w1 in adjList[u - 1]:
                if v in visit:
                    continue

                heapq.heappush(h, [w + w1, v])

        return time if len(visit) == n else -1


# Test helper
def test_network_delay_time():
    solution = Solution()

    # Test case 1: Example 1
    times1 = [[2, 1, 1], [2, 3, 1], [3, 4, 1]]
    n1 = 4
    k1 = 2
    expected1 = 2
    actual1 = solution.networkDelayTime(times1, n1, k1)
    print(f"{actual1} | {expected1}")

    # Test case 2: Example 2
    times2 = [[1, 2, 1]]
    n2 = 2
    k2 = 1
    expected2 = 1
    actual2 = solution.networkDelayTime(times2, n2, k2)
    print(f"{actual2} | {expected2}")

    # Test case 3: Example 3
    times3 = [[1, 2, 1]]
    n3 = 2
    k3 = 2
    expected3 = -1
    actual3 = solution.networkDelayTime(times3, n3, k3)
    print(f"{actual3} | {expected3}")

    # Test case 4: Single node
    times4 = []
    n4 = 1
    k4 = 1
    expected4 = 0
    actual4 = solution.networkDelayTime(times4, n4, k4)
    print(f"{actual4} | {expected4}")

    # Test case 5: Disconnected nodes
    times5 = [[1, 2, 1], [3, 4, 1]]
    n5 = 4
    k5 = 1
    expected5 = -1
    actual5 = solution.networkDelayTime(times5, n5, k5)
    print(f"{actual5} | {expected5}")

    # Test case 6: All connected
    times6 = [[1, 2, 1], [1, 3, 2], [2, 3, 1], [3, 4, 1]]
    n6 = 4
    k6 = 1
    expected6 = 3
    actual6 = solution.networkDelayTime(times6, n6, k6)
    print(f"{actual6} | {expected6}")


if __name__ == "__main__":
    test_network_delay_time()
